lib/streamlit/runtime/scriptrunner/script_runner.py

```diff
@@ -427,21 +429,47 @@ def _run_script(self, rerun_data: RerunData) -> None:
                 # in a 404 should the user click on them.
                 runtime.get_instance().media_file_mgr.clear_session_refs()
 
-            self._pages_manager.set_script_intent(
-                rerun_data.page_script_hash, rerun_data.page_name
-            )
-            active_script = self._pages_manager.get_initial_active_script(
-                rerun_data.page_script_hash, rerun_data.page_name
-            )
-            main_page_info = self._pages_manager.get_main_page()
+            ctx = self._get_script_run_ctx()
 
-            page_script_hash = (
-                active_script["page_script_hash"]
-                if active_script is not None
-                else main_page_info["page_script_hash"]
-            )
+            intended_page_script_hash = rerun_data.page_script_hash
+            intended_page_name = rerun_data.page_name
+            main_page = self._pages_manager.get_main_page()
+            # main_script_hash = main_page["page_script_hash"]
 
-            ctx = self._get_script_run_ctx()
+            # We want to know:
+            # 1. The active script to start with for the run.
+            # 2. The page_script_hash that we're running.
+
+            active_script = None
+            page_script_hash = None
+            if self._pages_manager.mpa_version == 1:
+                active_script = self._pages_manager.find_page_info(
+                    intended_page_script_hash, intended_page_name
+                )
+
+                if active_script is None:
+                    # At this point, we know that either
+                    #   * the script corresponding to the hash requested no longer
+                    #     exists, or
+                    #   * we were not able to find a script with the requested page
+                    #     name.
+                    # In both of these cases, we want to send a page_not_found
+                    # message to the frontend.
+                    msg = ForwardMsg()
+                    msg.page_not_found.page_name = intended_page_name
+                    ctx.enqueue(msg)
```

Comment:
Why can this only happen in MPAv1?

Reply:
Because the navigation call in MPA v2 will handle the page not found when called.